File: src/client/java/net/lixir/vminus/client/datagen/sound/SoundDefinitionProvider.java
package net.lixir.vminus.client.datagen.sound;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonArray;
import com.google.gson.JsonObject;
import net.minecraft.data.DataOutput;
import net.minecraft.data.DataProvider;
import net.minecraft.data.DataWriter;
import org.jetbrains.annotations.NotNull;

import java.util.List;
import java.util.concurrent.CompletableFuture;

public class SoundDefinitionProvider implements DataProvider {
    private static final Gson GSON = new GsonBuilder().setPrettyPrinting().create();
    private final DataOutput output;
    private final String modId;

    public SoundDefinitionProvider(DataOutput output, String modId) {
        this.output = output;
        this.modId = modId;
    }

    @Override
    public @NotNull CompletableFuture<?> run(@NotNull DataWriter writer) {
        JsonObject soundsJson = new JsonObject();

        /*
        for (SoundDefinitionInfo info : VRegistry.fromId(modId).getSoundDefinitionInfo()) {
            soundsJson.add(info.getSoundEventId(), generateSoundEntry(info));
        }


        Path target = output.getResolver(DataOutput.OutputType.RESOURCE_PACK, modId)
                .resolve(new Identifier(modId, "sounds"), ".json");

        try {
            Files.createDirectories(target.getParent());
        } catch (IOException e) {
            throw new RuntimeException("Failed to create directories for sounds.json", e);
        }

        return DataProvider.writeToPath(writer, soundsJson, target);

         */
        return null;
    }

    private JsonObject generateSoundEntry(SoundDefinitionInfo info) {
        JsonObject entry = new JsonObject();

        // key
        if (info.getSubtitle() != null) {
            String subtitle = info.getSubtitle().equals("default") ? "subtitles." + info.getPath() : info.getSubtitle();
            entry.addProperty("key", subtitle);
        }

        // sounds array
        JsonArray soundsArray = new JsonArray();
        List<String> paths = info.getPaths();
        int count = info.getCount();
        boolean isOpus = paths != null && !paths.isEmpty() && paths.get(0).endsWith(".opus");

        if (paths != null && !paths.isEmpty()) {
            for (String path : paths) {
                JsonObject soundObj = new JsonObject();
                soundObj.addProperty("name", path);
                soundsArray.add(soundObj);
            }
        } else {
            String base = info.getPath();
            for (int i = 1; i <= count; i++) {
                String file = base + (count > 1 ? i : "") + (isOpus ? ".opus" : ".ogg");
                JsonObject soundObj = new JsonObject();
                soundObj.addProperty("name", modId + ":" + file);
                soundsArray.add(soundObj);
            }
        }

        entry.add("sounds", soundsArray);
        return entry;
    }

    @Override
    public @NotNull String getName() {
        return "SoundDefinitionProvider [" + modId + "]";
    }
}

File: src/client/kotlin/net/lixir/vminus/client/datagen/provider/SoundDefinitionProvider.kt
package net.lixir.vminus.client.datagen.provider

import com.google.gson.JsonArray
import com.google.gson.JsonObject
import lombok.EqualsAndHashCode
import lombok.ToString
import net.fabricmc.fabric.api.datagen.v1.FabricDataOutput
import net.lixir.vminus.Vminus.Companion.devLogger
import net.minecraft.data.DataProvider
import net.minecraft.data.DataWriter
import net.minecraft.sound.SoundEvent
import net.minecraft.util.Identifier
import org.apache.logging.log4j.LogManager
import org.apache.logging.log4j.Logger
import java.util.concurrent.CompletableFuture

abstract class SoundDefinitionProvider(protected val output: FabricDataOutput) : DataProvider {
    private val modId: String = output.modId
    private val soundEntries: MutableMap<String, SoundEntry> = LinkedHashMap()

    protected abstract fun generateSounds(output: FabricDataOutput?)

    protected fun add(
        event: SoundEvent,
        subtitle: String? = null,
        replace: Boolean = false,
        count: Int = 1,
        vararg soundLocations: Any // either String paths or SoundVariant
    ): SoundEntry {
        val entry = SoundEntry(subtitle, replace)
        for (loc in soundLocations) {
            when (loc) {
                is String -> {
                    for (i in 1..count) {
                        var finalName = if (count > 1) appendIndexBeforeExtension(loc, i) else loc
                        if (!finalName.contains(":")) finalName = "$modId:$finalName"
                        entry.sounds.add(SoundVariant(Identifier.tryParse(finalName)))
                    }
                }

                is SoundVariant -> entry.sounds.add(loc)
                else -> throw IllegalArgumentException("Unsupported sound variant type: $loc")
            }
        }

        soundEntries[event.id.path] = entry
        return entry
    }

    override fun run(writer: DataWriter): CompletableFuture<*> {
        generateSounds(output)
        val root = JsonObject()
        for ((key, value) in soundEntries) {
            devLogger("Generating sound entry $key of $value")
            root.add(key, value.toJson())
        }

        val path = output.path.resolve("assets").resolve(modId).resolve("sounds.json")
        devLogger("Writing sounds.json to: ${path.toAbsolutePath()}")

        return DataProvider.writeToPath(writer, root, path)
    }

    override fun getName(): String = "Sound Definitions"

    class SoundEntry(
        private var subtitle: String? = null,
        private var replace: Boolean = false,
        val sounds: MutableList<SoundVariant> = ArrayList()
    ) {
        fun toJson(): JsonObject {
            val obj = JsonObject()
            subtitle?.let { obj.addProperty("key", it) }
            if (replace) obj.addProperty("replace", true)
            val arr = JsonArray()
            sounds.forEach { arr.add(it.toJson()) }
            obj.add("sounds", arr)
            return obj
        }
    }


    class SoundVariant(val identifier: Identifier?) {
        private var stream: Boolean = false
        private var volume: Float = 1.0f
        private var pitch: Float = 1.0f

        fun toJson(): JsonObject {
            val obj = JsonObject()
            obj.addProperty("name", identifier.toString())
            if (stream) obj.addProperty("stream", true)
            if (volume != 1.0f) obj.addProperty("volume", volume)
            if (pitch != 1.0f) obj.addProperty("pitch", pitch)
            return obj
        }
    }

    companion object {
        private fun appendIndexBeforeExtension(path: String, index: Int): String {
            val dotIndex = path.lastIndexOf('.')
            return if (dotIndex == -1 || dotIndex == path.length - 1) {
                path + index
            } else {
                path.substring(0, dotIndex) + index + path.substring(dotIndex)
            }
        }
    }
}

